xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp

```diff
@@ -561,30 +561,15 @@ void CGUIDialogPVRTimerSettings::OnSettingAction(const std::shared_ptr<const CSe
 
 bool CGUIDialogPVRTimerSettings::Validate()
 {
-  bool bStartAnyTime = m_bStartAnyTime;
-  bool bEndAnyTime = m_bEndAnyTime;
-
-  if (!m_timerType->SupportsStartAnyTime() ||
-      !m_timerType->IsEpgBased()) // Start anytime toggle is not displayed
-    bStartAnyTime = false; // Assume start time change needs checking for
-
-  if (!m_timerType->SupportsEndAnyTime() ||
-      !m_timerType->IsEpgBased()) // End anytime toggle is not displayed
-    bEndAnyTime = false; // Assume end time change needs checking for
-
-  // Begin and end time
-  if (!bStartAnyTime && !bEndAnyTime)
+  // When timer has a start and end time, compare them but only when it's
+  // not a timer rule, since those only have times not dates
+  if (!m_timerType->IsTimerRule() && !m_bStartAnyTime && !m_bEndAnyTime &&
+      m_timerType->SupportsStartTime() && m_timerType->SupportsEndTime() &&
```

Comment:
The main difference is the original code has a requirement for having both times or no times at all. The original code was for manual timers only.
```
(m_timerType->SupportsStartTime() == m_timerType->SupportsEndTime()) &&
```

You code is easier to read and I'm not sure the case where neither start or end time is displayed is valid, making your code more correct.

Reply:
Your code seems to apply to timer-rules, where mine applies to manual timers. So was the original always wrong then?

Furthermore, please consider my previous comment: "Well I did some more testing here with both normal (epg) timers and (epg) timer rules and it seems for the TVHeadend backend there's no point in validating anything at all. For TVHeadend timer rules at most it can have a start and end time (or start time only), but you can't check those since it doesn't have a date so a start of 23:00 and end of 1:00 shouldn't fail. And for normal timers with a date it's impossible to change the date spiners so that endTimeDate < startTimeDate." 

So I'm still wondering: do we need Validate() at all? Perhaps other code was added that made it useless? Or is it simply only useful with certain pvr backends (!=tvheadend)?

Reply:
If you read the original two PRs you’ll see that is not the case, this was all about manual timers. There was quite a bit of work that went into being able to have a validate function. I don’t use TVHeadend so i can’t confirm your assumption. 

If we use TVHeadend as the benchmark I’m sure we can simplify a lot of things but we need to be careful not to break functionality for the rest of the addons.

Reply:
The best approach here is likely to find the minimal change that allows the bug you found to be fixed so we can lessen any inspect on other addons.

Reply:
I finally found the use case this original code was for!

https://github.com/xbmc/xbmc/issues/19548. So I don't think your code with satisfy this case.

Reply:
> I finally found the use case this original code was for!
> 
> #19548. So I don't think your code with satisfy this case.

I think it does :-) The validation will only occur when the timer supports start/end-times and #19548 explicitly mentions an issue where validation is performed when no start/end times can be set. 

I still feel like the logic I've implemented now makes the most sense and as a bonus it's easy to read/straightforward. Wouldn't it be better to use this as a new starting-point since the old code was unreadable (and was causing regressions). If we keep using the old code there maybe other regressions that we simply don't know.... Agreed?

Reply:
> If you read the original two PRs you’ll see that is not the case, this was all about manual timers. There was quite a bit of work that went into being able to have a validate function. I don’t use TVHeadend so i can’t confirm your assumption.
> 
> If we use TVHeadend as the benchmark I’m sure we can simplify a lot of things but we need to be careful not to break functionality for the rest of the addons.

Well, in my case the code is triggered with timer-rules which is caused by that " if (!(m_timerType->SupportsStartTime() == m_timerType->SupportsEndTime() && m_timerType->IsTimerRule()) ...." always evaluates to true for timer-rules.

So if really, really want to fix the original code it, imo, should become " if (!(m_timerType->SupportsStartTime() == m_timerType->SupportsEndTime()) && !m_timerType->IsTimerRule() ..." . I suspect that's also what you intended back then but you placed the closing bracket in the wrong position... But I'm still not 100% sure about the rest, so I'd still prefer my proposed change....

Reply:
@phunkyfish : Heads up about #19548 which is ALSO about problems/regressions with specifically timer-rules. So I'm really starting to believe the Validate() code should only kick in for timer-rules to handle specific pvr backends. But it looks like HDHomerun acts a lot like TVHeadend concerning timer-rules.

Reply:
> > If you read the original two PRs you’ll see that is not the case, this was all about manual timers. There was quite a bit of work that went into being able to have a validate function. I don’t use TVHeadend so i can’t confirm your assumption.
> > If we use TVHeadend as the benchmark I’m sure we can simplify a lot of things but we need to be careful not to break functionality for the rest of the addons.
> 
> Well, in my case the code is triggered with timer-rules which is caused by that " if (!(m_timerType->SupportsStartTime() == m_timerType->SupportsEndTime() && m_timerType->IsTimerRule()) ...." always evaluates to true for timer-rules.

How exactly? The `!` negates what is inside, so it would never evaluate to true for a timer rule.

Reply:
> > > If you read the original two PRs you’ll see that is not the case, this was all about manual timers. There was quite a bit of work that went into being able to have a validate function. I don’t use TVHeadend so i can’t confirm your assumption.
> > > If we use TVHeadend as the benchmark I’m sure we can simplify a lot of things but we need to be careful not to break functionality for the rest of the addons.
> > 
> > 
> > Well, in my case the code is triggered with timer-rules which is caused by that " if (!(m_timerType->SupportsStartTime() == m_timerType->SupportsEndTime() && m_timerType->IsTimerRule()) ...." always evaluates to true for timer-rules.
> 
> How exactly? The `!` negates what is inside, so it would never evaluate to true for a timer rule.

Yes it does but the unreadable way (no offense) it is now hides it. Guide-based timer-rules with (at least) TVHeadend support a starttime but NOT an endtime. Making the evaluation: "if !(1 == 0 && 1)" which (obviously) results in "true". That's why I'm starting to believe that the fix in 39118772140b5835615afae56f35286c9b06629c is wrong or at least caused a regression with TVHeadend.